chain/chain/src/chain.rs

```diff
@@ -284,7 +284,24 @@ impl Chain {
     pub fn save_block(&mut self, block: &Block) -> Result<(), Error> {
         let mut chain_store_update = ChainStoreUpdate::new(&mut self.store);
 
-        // Some basic validation is needed here, #1236 is tracking it
+        // TODO check the description below
+        // Before saving the very first block in the epoch,
+        // we ensure that the block contains of correct data.
+        // It can be proven because hash of the block which
+        // identifies the block includes that data.
+
+        // 1. Checking state_root validity
+        let state_root = Block::compute_state_root(&block.chunks);
+        if block.header.inner.prev_state_root != state_root {
+            return Err(ErrorKind::InvalidStateRoot.into());
+        }
+
+        // 2. Checking tx_root validity
+        let tx_root = Block::compute_tx_root(&block.transactions);
+        if block.header.inner.tx_root != tx_root {
+            return Err(ErrorKind::InvalidTxRoot.into());
+        }
+
```

Comment:
1. I think we don't need to change the way of computing the hash of the block because currently the hash of the block is the hash of the header of the block, and header includes hashes of state (chunks) and txs. That's why I can say that hash of the header represents the block.
2. I rely on NetworkClientMessages::Block which calls save_block. It already checks that block.hash() (which is hash of the header, by implementation) is exactly what we expect to receive.

Despite I wrote above, I agreed with you, it's better to put all the checks inside fn save_block.